src/git/negociator.ml

```diff
@@ -0,0 +1,413 @@
+               (for any reason), we just avoid the rest of the compute (so we
+               avoid the rest of the ancestors).
+
+               For me, we need to notice to the user the error (so avoid the
+               next compute and return only the error). So we need to think
+               about that if it's what we really want. TODO! *)
+            | Ok value ->
+              mark_common t bucket ~ancestors:false hash value rev >>= fun rev -> go rev rest
+        in
+
+        let non_common_revs =
+          if not ancestors && not (Flag.is_popped value.V.flags)
+          then rev.non_common_revs - 1
+          else rev.non_common_revs
+        in
+
+        go { rev with non_common_revs; } (Store.Value.Commit.parents value.V.commit)
+      end
+    end else Lwt.return rev
+
+  let get t bucket rev =
+    let open Lwt.Infix in
+
+    let rec go rev =
+      if rev.non_common_revs = 0
+      then Lwt.return (None, rev)
+      else match Pq.pop rev.pq with
+      | None -> Lwt.return (None, rev)
+      | Some ((hash, ({ V.commit; flags; } as value)), pq) ->
+        value.V.flags <- Flag.to_popped flags;
+
+        let non_common_revs =
+          if not (Flag.is_common flags)
+          then rev.non_common_revs - 1
+          else rev.non_common_revs
+        in
+
+        let leave, marks =
+          if Flag.is_common flags
+          then false, Flag.to_common @@ Flag.to_seen @@ Flag.default
+          (* XXX(dinosaure): do not send "have", and ignore ancestors. *)
+          else if Flag.is_common_ref flags
+          then true,  Flag.to_common @@ Flag.to_seen @@ Flag.default
+          (* XXX(dinosaure): send "have" and ignore ancestors. *)
+          else true,  Flag.to_seen @@ Flag.default
+          (* XXX(dinosaure): send "have", also for its ancestors. *)
+        in
+
+        let rec gogo rev = function
+          | [] -> Lwt.return rev
+          | hash :: rest ->
+            Bucket.get t bucket hash >>= function
+            | Ok ({ V.commit; flags; } as value) ->
+              let rev =
+                if not (Flag.is_seen flags)
+                then push hash value marks rev
+                else rev
+              in
+
+              (if Flag.is_common marks
+               then mark_common t bucket ~ancestors:true hash value rev
+               else Lwt.return rev)
+              >>= fun rev -> gogo rev rest
+            | Error _ -> Lwt.return rev
+        in
+
+        gogo { pq; non_common_revs; } (Store.Value.Commit.parents commit) >>= fun rev ->
+
+        if leave
+        then Lwt.return  (Some hash, rev)
+        else go rev
+    in
+
+    go rev
+
+  type state =
+    { ready    : bool (* got ready detail *)
+    ; continue : bool (* got continue detail *)
+    ; finish   : bool (* want to finish *)
+    ; count    : int  (* how many hashes we get *)
+    ; flush    : int  (* how many hashes before flush *)
+    ; vain     : int  (* how many vainly hashes we have *)
```

Comment:
I can change the value of `_VAIN` (by functor), I put 128 to test some complex computation between a client and a server.